### DISCUSSION TITLE
Support depot.dev or alternatives to accelerate builds on M1

### DISCUSSION THREAD

**Discussion by MikeBinstock** [basecamp/kamal#309] [General]
Yesterday I learn't about depot.dev, it's a build on demand service to accelerate build times. They have a free plan. 
It would be great if we can integrate this on MRSK.

**Comment by dhh** [comment]
Would be happy to see someone explore this. Could be part of a full adapterization of the docker command structure. This would pave the way for podman as well. Please do investigate.

**Comment by stormmaster42** [comment]
Looking into this - as we are also using `depot.dev` to speed up our image build and I want to move to kamal for newer services.
kamal and depot both solve the target architecture problem. That is why depot does not work out of the box with depot.
depot initializes itself as a docker cli plugin and sets itself as default build handler.

Looking at `Kamal::Commands::Builder` and Kamal::Commands::Builder::Native:
The native builder is closest to what we need to get depot working, as it is using docker without defining its own buildx extensions.

Actually kamal can be _tricked_ to work with depot with following configuration for the builder (running this on a M1):
```
builder:
  local:
    arch: amd64
  # settings below force build to use depot.dev
  multiarch: false
  args:
    - "A=1 --load"
```
This forces kamal to do a native build, which in turn is then using the depot plugin (`depot init` before). The `--load` is necessary, because depot does not pull the image back to the kamal runner and kamal in the native builder pushes the image to the registry. With depot this would be done with an additional `--push`.

The simplest way to move forward here, would be to allow to override the build command to be used with a config like:
```
builder:
  target: depot
```
And provide a targeted implementation (plugin?) as `Kamal::Commands::Builder::<target>`

The command for depot would look something like this:
```
class Kamal::Commands::Builder::Debot < Kamal::Commands::Builder::Native
  def push
    combine \
      docker(:build, *build_options, build_context),
      docker(:push, config.absolute_image),
      docker(:push, config.latest_image)
  end

  def build_options
    super + ['--push']
  end
end
```

Only change to main code base would then be in `Kamal::Commands::Builder#target` to pick up that new config?

Does that sound like a plan?

**Comment by navinpeiris** [comment]
For depot.dev support specifically, there's a fork of Kamal here with depot integration: https://github.com/navinpeiris/kamal-depot

**Comment by jtarchie** [comment]
@navinpeiris, did you have plans to PR this?

**Reply by navinpeiris** [reply]
Heya JT, yes I did, but that was closed in favour of exploring a plugin based implementation. You can see it here: https://github.com/basecamp/kamal/pull/394
